### PR TITLE
Create data source for organization ip allow list

### PR DESCRIPTION
**Pull request by douglascayers**
### Summary

Allow terraform to retrieve an enterprise cloud organization's [IP allow list](https://docs.github.com/en/enterprise-cloud@latest/organizations/keeping-your-organization-secure/managing-security-settings-for-your-organization/managing-allowed-ip-addresses-for-your-organization).

This data source is a precursor to a future PR to add a resource to allow terraform to manage IP allow list entries as requested in https://github.com/integrations/terraform-provider-github/issues/1067.

### Changes

* Add data source to retrieve an organization's IP allow list via APIv4 (graphql)
* Paginates 100 at a time to retrieve all entries
* Add unit test and documentation

### Tests

* `github/data_source_github_organization_ip_allow_list_test.go`

### Usage

Example configuration:
```hcl
data "github_organization_ip_allow_list" "all" {}
```

Example output:
```json
{
  "resources": [
    {
      "mode": "data",
      "type": "github_organization_ip_allow_list",
      "name": "all",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "id": "MDEyOk9yZ2FuaXphdGlvbjg0NDkwOTE4",
            "ip_allow_list": [
              {
                "allow_list_value": "192.30.252.0/22",
                "created_at": "2022-06-10T22:58:22Z",
                "id": "IALE_kwHOBQk6ps4ACyRH",
                "is_active": true,
                "name": "GitHub Hooks",
                "updated_at": "2022-06-10T22:58:22Z"
              },
              {
                "allow_list_value": "185.199.108.0/22",
                "created_at": "2022-06-10T22:58:22Z",
                "id": "IALE_kwHOBQk6ps4ACyRH",
                "is_active": true,
                "name": "GitHub Hooks",
                "updated_at": "2022-06-10T22:58:22Z"
              }
            ]
          },
          "sensitive_attributes": []
        }
      ]
    }
  ]
}
```

### CODE REVIEW

**Comment by douglascayers** [comment]
Hi @kfcampbell, would appreciate your review of this PR to add data source to retrieve an org's IP allow list. Thanks!

File: github/provider.go [hunk 1]
```diff
@@ -134,8 +134,9 @@ func Provider() terraform.ResourceProvider {
 		},
 
 		DataSourcesMap: map[string]*schema.Resource{
-			"github_actions_secrets":                 dataSourceGithubActionsSecrets(),
+			"github_actions_organization_secrets":    dataSourceGithubActionsOrganizationSecrets(),
```
Comment (douglascayers) [inline github/provider.go:137]: sorted the resources

File: github/provider.go [hunk 2]
```diff
@@ -145,7 +146,7 @@ func Provider() terraform.ResourceProvider {
 			"github_ip_ranges":                       dataSourceGithubIpRanges(),
 			"github_membership":                      dataSourceGithubMembership(),
 			"github_organization":                    dataSourceGithubOrganization(),
-			"github_actions_organization_secrets":    dataSourceGithubActionsOrganizationSecrets(),
+			"github_organization_ip_allow_list":      dataSourceGithubOrganizationIpAllowList(),
```
Comment (douglascayers) [inline github/provider.go:149]: sorted the resources by moving actions resource to top; then added new the ip allow list resource

**Comment by kfcampbell** [comment]
I'm seeing the following error when running the integration test locally: 

````
    testing.go:705: Step 0 error: Check failed: 6 errors occurred:
        	* Check 1/6 error: data.github_organization_ip_allow_list.all: Attribute 'ip_allow_list.0.id' expected to be set
        	* Check 2/6 error: data.github_organization_ip_allow_list.all: Attribute 'ip_allow_list.0.name' expected to be set
        	* Check 3/6 error: data.github_organization_ip_allow_list.all: Attribute 'ip_allow_list.0.allow_list_value' expected to be set
        	* Check 4/6 error: data.github_organization_ip_allow_list.all: Attribute 'ip_allow_list.0.is_active' expected to be set
        	* Check 5/6 error: data.github_organization_ip_allow_list.all: Attribute 'ip_allow_list.0.created_at' expected to be set
        	* Check 6/6 error: data.github_organization_ip_allow_list.all: Attribute 'ip_allow_list.0.updated_at' expected to be set
```

and the test suite doesn't pass. Is it succeeding for you locally?

**Comment by douglascayers** [comment]
Hrmm, will take a look. It was passing for sure before I back merged with main branch today. I saw the commit check pass so thought that meant the back merge went well.

**Comment by douglascayers** [comment]
@kfcampbell The test succeeds locally for me. How are you running the test?

This is what I do from the root of the project:
```
$ go test -run TestAccGithubOrganizationIpAllowListDataSource ./github
ok      github.com/integrations/terraform-provider-github/v5/github     0.494s
```

**Comment by douglascayers** [comment]
Oh, I just noticed the key part in that test result:
```
testing.go:567: Acceptance tests skipped unless env 'TF_ACC' set
```

Stand by, let me set some vars and retest to confirm ⏳

**Comment by douglascayers** [comment]
Yes, that test passes for me locally.

```shell
$ make build
==> Checking that code complies with gofmt requirements...
go build ./...

$ export GITHUB_TOKEN="ghp_<my_token>"
$ export GITHUB_ORGANIZATION="<my_org>"

$ TF_ACC=1 go test -v -run TestAccGithubOrganizationIpAllowListDataSource ./github
=== RUN   TestAccGithubOrganizationIpAllowListDataSource
=== RUN   TestAccGithubOrganizationIpAllowListDataSource/queries_without_error
=== RUN   TestAccGithubOrganizationIpAllowListDataSource/queries_without_error/with_an_anonymous_account
    data_source_github_organization_ip_allow_list_test.go:40: anonymous account not supported for this operation
=== RUN   TestAccGithubOrganizationIpAllowListDataSource/queries_without_error/with_an_individual_account
    data_source_github_organization_ip_allow_list_test.go:44: individual account not supported for this operation
=== RUN   TestAccGithubOrganizationIpAllowListDataSource/queries_without_error/with_an_organization_account
--- PASS: TestAccGithubOrganizationIpAllowListDataSource (17.58s)
    --- PASS: TestAccGithubOrganizationIpAllowListDataSource/queries_without_error (17.58s)
        --- SKIP: TestAccGithubOrganizationIpAllowListDataSource/queries_without_error/with_an_anonymous_account (0.00s)
        --- SKIP: TestAccGithubOrganizationIpAllowListDataSource/queries_without_error/with_an_individual_account (0.00s)
        --- PASS: TestAccGithubOrganizationIpAllowListDataSource/queries_without_error/with_an_organization_account (17.58s)
PASS
ok      github.com/integrations/terraform-provider-github/v5/github     18.058s
```

**Comment by kfcampbell** [comment]
Unfortunately, our integration tests are being run manually at this stage and are not automatically run on PR checks. 

This is weird...I have different results. When I'm up-to-date with your branch, I set the following environment variables:

```bash
 sh$ env | grep GITHUB
GITHUB_TOKEN=REDACTED
GITHUB_ORGANIZATION=kfcampbell-terraform-provider

 sh$ env | grep TF_
TF_ACC=1
```

Then I'll run the test using your same command: 

```
 sh$ TF_ACC=1 go test -v -run TestAccGithubOrganizationIpAllowListDataSource ./github
=== RUN   TestAccGithubOrganizationIpAllowListDataSource
=== RUN   TestAccGithubOrganizationIpAllowListDataSource/queries_without_error
=== RUN   TestAccGithubOrganizationIpAllowListDataSource/queries_without_error/with_an_anonymous_account
    data_source_github_organization_ip_allow_list_test.go:40: anonymous account not supported for this operation
=== RUN   TestAccGithubOrganizationIpAllowListDataSource/queries_without_error/with_an_individual_account
    data_source_github_organization_ip_allow_list_test.go:44: individual account not supported for this operation
=== RUN   TestAccGithubOrganizationIpAllowListDataSource/queries_without_error/with_an_organization_account
    testing.go:705: Step 0 error: Check failed: 6 errors occurred:
        	* Check 1/6 error: data.github_organization_ip_allow_list.all: Attribute 'ip_allow_list.0.id' expected to be set
        	* Check 2/6 error: data.github_organization_ip_allow_list.all: Attribute 'ip_allow_list.0.name' expected to be set
        	* Check 3/6 error: data.github_organization_ip_allow_list.all: Attribute 'ip_allow_list.0.allow_list_value' expected to be set
        	* Check 4/6 error: data.github_organization_ip_allow_list.all: Attribute 'ip_allow_list.0.is_active' expected to be set
        	* Check 5/6 error: data.github_organization_ip_allow_list.all: Attribute 'ip_allow_list.0.created_at' expected to be set
        	* Check 6/6 error: data.github_organization_ip_allow_list.all: Attribute 'ip_allow_list.0.updated_at' expected to be set


--- FAIL: TestAccGithubOrganizationIpAllowListDataSource (2.32s)
    --- FAIL: TestAccGithubOrganizationIpAllowListDataSource/queries_without_error (2.32s)
        --- SKIP: TestAccGithubOrganizationIpAllowListDataSource/queries_without_error/with_an_anonymous_account (0.00s)
        --- SKIP: TestAccGithubOrganizationIpAllowListDataSource/queries_without_error/with_an_individual_account (0.00s)
        --- FAIL: TestAccGithubOrganizationIpAllowListDataSource/queries_without_error/with_an_organization_account (2.32s)
FAIL
FAIL	github.com/integrations/terraform-provider-github/v5/github	2.333s
FAIL
```

FWIW, I have never set or modified IP allowlists for my test organization, which may be why this is failing.

**Comment by kfcampbell** [comment]
Yep, manually setting an IP and enabling an allowlist organization makes the test pass. 

What would you prefer to do here? We could merge as-is with the idea that once #1067 is closed, we could come back and fix the test.

**Comment by douglascayers** [comment]
@kfcampbell Thanks for confirming.

Hrm, I don't have an Enterprise Org that I can test where it _doesn't_ have IP allow lists added but the feature is enabled.

Does the test fail for you if the feature is enabled but no IPs allowed? Or fail only if the feature isn't enabled?

If it fails only when the feature isn't enabled then perhaps the test can become a bit smarter to skip orgs with a message that the feature isn't enabled, can't test. But if the feature is enabled then it'll do the call?

Or do you think the "fix" is simpler, just check for an empty list coming back?

**Comment by douglascayers** [comment]
@kfcampbell you know what, I don't think an org can both have IP allow list on AND have no IPs listed as that would prevent the users from accessing it entirely. So there likely has to be at least one IP. 

In that case, I think this is good to merge as-is. With the known room for improvement that the test should skip if the feature isn't enabled to explain why the test didn't run.

**Comment by kfcampbell** [comment]
😄 That works for me! I'll merge and release this, and I'm looking forward to checking out any progress on #1067.

**Review by kfcampbell — APPROVED** [review]
After #1067 is closed, the test `TestAccGithubOrganizationIpAllowListDataSource` should be amended to ensure that the organization being tested has the IP allowlist feature enabled.

**Pull request merged**